## Libraries

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Analyse exploratoire

In [4]:
df = pd.read_csv("vehicles.csv")

In [5]:
# Afficher les cinq premières lignes du df
df.head()

id                                                url  \
0  7222695916  https://prescott.craigslist.org/cto/d/prescott...   
1  7218891961  https://fayar.craigslist.org/ctd/d/bentonville...   
2  7221797935  https://keys.craigslist.org/cto/d/summerland-k...   
3  7222270760  https://worcester.craigslist.org/cto/d/west-br...   
4  7210384030  https://greensboro.craigslist.org/cto/d/trinit...   

                   region                         region_url  price  year  \
0                prescott    https://prescott.craigslist.org   6000   NaN   
1            fayetteville       https://fayar.craigslist.org  11900   NaN   
2            florida keys        https://keys.craigslist.org  21000   NaN   
3  worcester / central MA   https://worcester.craigslist.org   1500   NaN   
4              greensboro  https://greensboro.craigslist.org   4900   NaN   

  manufacturer model condition cylinders  ... size  type paint_color  \
0          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
1          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
2          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
3          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
4          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   

  image_url description county state lat long posting_date  
0       NaN         NaN    NaN    az NaN  NaN          NaN  
1       NaN         NaN    NaN    ar NaN  NaN          NaN  
2       NaN         NaN    NaN    fl NaN  NaN          NaN  
3       NaN         NaN    NaN    ma NaN  NaN          NaN  
4       NaN         NaN    NaN    nc NaN  NaN          NaN  

[5 rows x 26 columns]

In [6]:
# Afficher les cinq dernières lignes du df
df.tail()

id                                                url  \
426875  7301591192  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426876  7301591187  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426877  7301591147  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426878  7301591140  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426879  7301591129  https://wyoming.craigslist.org/ctd/d/atlanta-2...   

         region                      region_url  price    year manufacturer  \
426875  wyoming  https://wyoming.craigslist.org  23590  2019.0       nissan   
426876  wyoming  https://wyoming.craigslist.org  30590  2020.0        volvo   
426877  wyoming  https://wyoming.craigslist.org  34990  2020.0     cadillac   
426878  wyoming  https://wyoming.craigslist.org  28990  2018.0        lexus   
426879  wyoming  https://wyoming.craigslist.org  30590  2019.0          bmw   

                           model condition    cylinders  ... size       type  \
426875         maxima s sedan 4d      good  6 cylinders  ...  NaN      sedan   
426876  s60 t5 momentum sedan 4d      good          NaN  ...  NaN      sedan   
426877          xt4 sport suv 4d      good          NaN  ...  NaN  hatchback   
426878           es 350 sedan 4d      good  6 cylinders  ...  NaN      sedan   
426879  4 series 430i gran coupe      good          NaN  ...  NaN      coupe   

       paint_color                                          image_url  \
426875         NaN  https://images.craigslist.org/00o0o_iiraFnHg8q...   
426876         red  https://images.craigslist.org/00x0x_15sbgnxCIS...   
426877       white  https://images.craigslist.org/00L0L_farM7bxnxR...   
426878      silver  https://images.craigslist.org/00z0z_bKnIVGLkDT...   
426879         NaN  https://images.craigslist.org/00Y0Y_lEUocjyRxa...   

                                              description county state  \
426875  Carvana is the safer way to buy a car During t...    NaN    wy   
426876  Carvana is the safer way to buy a car During t...    NaN    wy   
426877  Carvana is the safer way to buy a car During t...    NaN    wy   
426878  Carvana is the safer way to buy a car During t...    NaN    wy   
426879  Carvana is the safer way to buy a car During t...    NaN    wy   

              lat       long              posting_date  
426875  33.786500 -84.445400  2021-04-04T03:21:31-0600  
426876  33.786500 -84.445400  2021-04-04T03:21:29-0600  
426877  33.779214 -84.411811  2021-04-04T03:21:17-0600  
426878  33.786500 -84.445400  2021-04-04T03:21:11-0600  
426879  33.779214 -84.411811  2021-04-04T03:21:07-0600  

[5 rows x 26 columns]

In [7]:
# Affciher les lignes du df, aléatoirement
df.sample(n=10)

id                                                url  \
408914  7316517917  https://seattle.craigslist.org/sno/ctd/d/lynnw...   
388062  7309720478  https://vermont.craigslist.org/cto/d/washingto...   
162729  7309834620  https://omaha.craigslist.org/ctd/d/omaha-2016-...   
210863  7302447251  https://bemidji.craigslist.org/ctd/d/grand-for...   
68596   7304931259  https://ventura.craigslist.org/cto/d/moorpark-...   
408063  7316808424  https://seattle.craigslist.org/tac/ctd/d/wilke...   
2298    7302411297  https://shoals.craigslist.org/cto/d/muscle-sho...   
258964  7303087737  https://southjersey.craigslist.org/ctd/d/vinel...   
164911  7315289478  https://kansascity.craigslist.org/ctd/d/indepe...   
13806   7303031598  https://sierravista.craigslist.org/ctd/d/tucso...   

                          region                          region_url  price  \
408914            seattle-tacoma      https://seattle.craigslist.org   5900   
388062                   vermont      https://vermont.craigslist.org  14500   
162729    omaha / council bluffs        https://omaha.craigslist.org  18995   
210863                   bemidji      https://bemidji.craigslist.org  11490   
68596             ventura county      https://ventura.craigslist.org   7900   
408063            seattle-tacoma      https://seattle.craigslist.org  48866   
2298    florence / muscle shoals       https://shoals.craigslist.org  16900   
258964              south jersey  https://southjersey.craigslist.org  19590   
164911           kansas city, MO   https://kansascity.craigslist.org   4995   
13806               sierra vista  https://sierravista.craigslist.org  11995   

          year   manufacturer                  model  condition    cylinders  \
408914  2007.0          honda              cr-v ex-l  excellent  4 cylinders   
388062  1981.0      chevrolet               corvette  excellent  8 cylinders   
162729  2016.0        lincoln                    mkc        NaN  4 cylinders   
210863  2013.0     volkswagen  beetle fender edition       good  4 cylinders   
68596   1997.0         jaguar                    xk6   like new  6 cylinders   
408063  2019.0      chevrolet     silverado 1500 rst       good  8 cylinders   
2298    2006.0           ford        f350 super duty        NaN          NaN   
258964  2016.0  mercedes-benz          e-class e 350       good  6 cylinders   
164911  2006.0           ford                 fusion        NaN          NaN   
13806   2014.0         nissan                 murano        NaN          NaN   

        ...       size       type paint_color  \
408914  ...        NaN        SUV         NaN   
388062  ...        NaN      coupe         NaN   
162729  ...        NaN        SUV       black   
210863  ...   mid-size  hatchback       black   
68596   ...  full-size      sedan       black   
408063  ...        NaN     pickup         NaN   
2298    ...        NaN        NaN         NaN   
258964  ...        NaN      sedan         NaN   
164911  ...        NaN      other        grey   
13806   ...        NaN        NaN         NaN   

                                                image_url  \
408914  https://images.craigslist.org/00e0e_jxwKelAWfq...   
388062  https://images.craigslist.org/00e0e_jQwpYm9rUZ...   
162729  https://images.craigslist.org/00606_da7noPSLbK...   
210863  https://images.craigslist.org/00o0o_dq25L0UPFL...   
68596   https://images.craigslist.org/00z0z_endrjf2mVL...   
408063  https://images.craigslist.org/01313_lHV9HEbJfl...   
2298    https://images.craigslist.org/00v0v_4jOf2Y3cf9...   
258964  https://images.craigslist.org/00O0O_dZbLRpmc0Z...   
164911  https://images.craigslist.org/01313_9TFk49heMX...   
13806   https://images.craigslist.org/00303_4ZN0GsnOjj...   

                                              description county state  \
408914  2007 Honda CR-V EX-L family owned with 208K mo...    NaN    wa   
388062  Very well documented.  Sales brochure from 198...    NaN    vt   
162729  2016 *Lincoln* *MKC

In [8]:
# Dimension du df c-a-d le nombre de lignes et colonnes
df.shape

(426880, 26)

In [9]:
# Nombre de colonnes dans df
df.columns

Index(['id', 'url', 'region', 'region_url', 'price', 'year', 'manufacturer',
       'model', 'condition', 'cylinders', 'fuel', 'odometer', 'title_status',
       'transmission', 'VIN', 'drive', 'size', 'type', 'paint_color',
       'image_url', 'description', 'county', 'state', 'lat', 'long',
       'posting_date'],
      dtype='object')

In [10]:
# Le type de données de chaque colonne du df
df.dtypes

id                int64
url              object
region           object
region_url       object
price             int64
year            float64
manufacturer     object
model            object
condition        object
cylinders        object
fuel             object
odometer        float64
title_status     object
transmission     object
VIN              object
drive            object
size             object
type             object
paint_color      object
image_url        object
description      object
county          float64
state            object
lat             float64
long            float64
posting_date     object
dtype: object

La commande ci-dessous, nous me permet d'avoir des informations sur le dataset comme le type de données et le nombre total des valeurs non nulles dans chaque colonne.

Cela nous permet didentifier les valeurs manquantes et les types de données incorrectes 

In [11]:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 26 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   url           426880 non-null  object 
 2   region        426880 non-null  object 
 3   region_url    426880 non-null  object 
 4   price         426880 non-null  int64  
 5   year          425675 non-null  float64
 6   manufacturer  409234 non-null  object 
 7   model         421603 non-null  object 
 8   condition     252776 non-null  object 
 9   cylinders     249202 non-null  object 
 10  fuel          423867 non-null  object 
 11  odometer      422480 non-null  float64
 12  title_status  418638 non-null  object 
 13  transmission  424324 non-null  object 
 14  VIN           265838 non-null  object 
 15  drive         296313 non-null  object 
 16  size          120519 non-null  object 
 17  type          334022 non-null  object 
 18  pain

In [12]:
#Statistiques descriptives pour les colonnes numériques (moyenne, écart-type, min, max, quartiles)
df.describe()

id         price           year      odometer  county  \
count  4.268800e+05  4.268800e+05  425675.000000  4.224800e+05     0.0   
mean   7.311487e+09  7.519903e+04    2011.235191  9.804333e+04     NaN   
std    4.473170e+06  1.218228e+07       9.452120  2.138815e+05     NaN   
min    7.207408e+09  0.000000e+00    1900.000000  0.000000e+00     NaN   
25%    7.308143e+09  5.900000e+03    2008.000000  3.770400e+04     NaN   
50%    7.312621e+09  1.395000e+04    2013.000000  8.554800e+04     NaN   
75%    7.315254e+09  2.648575e+04    2017.000000  1.335425e+05     NaN   
max    7.317101e+09  3.736929e+09    2022.000000  1.000000e+07     NaN   

                 lat           long  
count  420331.000000  420331.000000  
mean       38.493940     -94.748599  
std         5.841533      18.365462  
min       -84.122245    -159.827728  
25%        34.601900    -111.939847  
50%        39.150100     -88.432600  
75%        42.398900     -80.832039  
max        82.390818     173.885502

In [13]:
# Statistiques descriptives pour des colonnes catégorielles (nombre de valeurs uniques, fréquence des plus courantes).
df.describe(include='object')

url    region  \
count                                              426880    426880   
unique                                             426880       404   
top     https://prescott.craigslist.org/cto/d/prescott...  columbus   
freq                                                    1      3608   

                            region_url manufacturer   model condition  \
count                           426880       409234  421603    252776   
unique                             413           42   29667         6   
top     https://spokane.craigslist.org         ford   f-150      good   
freq                              2988        70985    8009    121456   

          cylinders    fuel title_status transmission                VIN  \
count        249202  423867       418638       424324             265838   
unique            8       5            6            3             118264   
top     6 cylinders     gas        clean    automatic  1FMJU1JT1HEA52352   
freq          94169  356209       405117       336524                261   

         drive       size    type paint_color  \
count   296313     120519  334022      296677   
unique       3          4      13          12   
top        4wd  full-size   sedan       white   
freq    131904      63465   87056       79285   

                                                image_url  \
count                                              426812   
unique                                             241899   
top     https://images.craigslist.org/00N0N_1xMPvfxRAI...   
freq                                                 7357   

                                              description   state  \
count                                              426810  426880   
unique                                             360911      51   
top     35 VEHICLES PRICED UNDER $3000!!!  BIG TIME! T...      ca   
freq                                                  231   50614   

                    posting_date  
count                     426812  
unique                    381536  
top     2021-04-23T22:13:05-0400  
freq                          12

## Nettoyage

In [14]:
df = pd.read_csv("vehicles.csv")

Cette partie consiste à manipuler le dataset dans le but de le nettoyer et de le traiter.

Pour ce faire, nous allons procéder en gérant les valeurs manquantes, doublons, abbérantes et corriger si nécessaire les types de données des colonnes non conformes



### Suppression des colonnes inutiles

In [15]:
df.drop(columns=['url','region_url','image_url','description'], inplace=True)

In [16]:
df.columns

Index(['id', 'region', 'price', 'year', 'manufacturer', 'model', 'condition',
       'cylinders', 'fuel', 'odometer', 'title_status', 'transmission', 'VIN',
       'drive', 'size', 'type', 'paint_color', 'county', 'state', 'lat',
       'long', 'posting_date'],
      dtype='object')

### Traitement des valeurs incohérentes

In [17]:
#Suppression des prix incohérents
df = df[(df['price'] >= 1000) & (df['price'] <= 250000)]
df.shape

(380470, 22)

In [18]:
#Suppression des années incohérentes
df = df[(df['year'] >= 1950) & (df['year'] <= 2025)]
df.shape

(378185, 22)

In [19]:
#Suppression des kilométrages incohérents
df = df[(df['odometer'] >= 100) & (df['odometer'] <= 500000)]
df.shape

(371031, 22)

### Gestion des valeurs manquantes

#### Suppresion des lignes avec des valeurs manquantes importantes

In [20]:
colonnes_importantes = ['price', 'year', 'manufacturer', 'model', 'odometer']
df.dropna(subset=colonnes_importantes, inplace=True)
df.shape

(354200, 22)

#### Remplir les valeurs manquantes dans les variables 'drive', 'cylinders' 'fuel' et 'transmission'

Pour eviter de perdre trop de lignes à cause d'un champ vide, Nous allons remplir les valeurs vident par la valeur majoritaire du modèle de voiture pour chaque variables.

In [ ]:
print("Valeurs manquantes restantes - drive :", df['drive'].isna().sum())
print("Valeurs manquantes restantes - cylinders :", df['cylinders'].isna().sum())
print("Valeurs manquantes restantes - fuel :", df['fuel'].isna().sum())
print("Valeurs manquantes restantes - transmission :", df['transmission'].isna().sum())

Valeurs manquantes restantes - drive : 79069
Valeurs manquantes restantes - cylinders : 105626
Valeurs manquantes restantes - fuel : 1564
Valeurs manquantes restantes - transmission : 1087


Nous commencons par identifier les valeurs majoritaires par modèle.

Pour ce faire, nous allons créer une série par variable qui va contenir la valeur la plus fréquente pour chaque modèle de voiture.

La ligne de code va regrouper les lignes par model, puis appliquer une agrégation pour donner la valeur la plus fréquente. 

Ensuite, lambda permet de prendre la première valeur pour éviter les blocages en cas d'égalité. Et pour finir, on vérifie qu'il y a bien une valeur dominante, sinon on garde NaN.

In [ ]:
# Pour la variable 'drive'
mode_drive_par_model = df.groupby('model')['drive'].agg(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

# Pour la variable 'cylinders'
mode_cyl_par_model = df.groupby('model')['cylinders'].agg(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

# Pour la variable 'fuel'
mode_cyl_par_model = df.groupby('model')['fuel'].agg(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

# Pour la variable 'transmission'
mode_cyl_par_model = df.groupby('model')['transmission'].agg(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

Dans cette deuxième étape, nous allons remplir les valeurs manquantes en se basant sur la valeur majoritaire obtenue avec le code précédent.

Nous avons appliqué une fonction ligne par ligne pour remplir les valeurs manquantes. Si une ligne avait une valeur manquante dans cette variable, et que le modèle de véhicule était présent dans notre dictionnaire des valeurs majoritaires, nous avons alors remplacé cette valeur par la valeur la plus fréquente observée pour ce modèle. 

Cette méthode permet une imputation ciblée, cohérente et basée sur les observations du dataset.

In [ ]:
# Remplir les valeurs manquantes de 'drive'
df['drive'] = df.apply(
    lambda row: mode_drive_par_model[row['model']] if pd.isna(row['drive']) and row['model'] in mode_drive_par_model else row['drive'],
    axis=1
)

# Remplir les valeurs manquantes de 'cylinders'
df['cylinders'] = df.apply(
    lambda row: mode_cyl_par_model[row['model']] if pd.isna(row['cylinders']) and row['model'] in mode_cyl_par_model else row['cylinders'],
    axis=1
)

# Remplir les valeurs manquantes de 'fuel'
df['fuel'] = df.apply(
    lambda row: mode_cyl_par_model[row['model']] if pd.isna(row['fuel']) and row['model'] in mode_cyl_par_model else row['fuel'],
    axis=1
)

# Remplir les valeurs manquantes de 'transmission'
df['transmission'] = df.apply(
    lambda row: mode_cyl_par_model[row['model']] if pd.isna(row['transmission']) and row['model'] in mode_cyl_par_model else row['transmission'],
    axis=1
)

In [35]:
print("Valeurs manquantes restantes - drive :", df['drive'].isna().sum())
print("Valeurs manquantes restantes - cylinders :", df['cylinders'].isna().sum())
print("Valeurs manquantes restantes - fuel :", df['fuel'].isna().sum())
print("Valeurs manquantes restantes - transmission :", df['transmission'].isna().sum())

Valeurs manquantes restantes - drive : 9331
Valeurs manquantes restantes - cylinders : 39
Valeurs manquantes restantes - fuel : 1
Valeurs manquantes restantes - transmission : 40


### Correction des types de données

In [23]:
# Conversion de la colonne 'posting_date' en format datetime sans les heures
df['posting_date'] = pd.to_datetime(df['posting_date'], errors='coerce', utc=True)
df['posting_date'] = df['posting_date'].dt.tz_convert(None).dt.normalize()

In [24]:
df.sample(n=10)

id          region  price    year   manufacturer  \
422586  7315508213       milwaukee  37000  2014.0      chevrolet   
269184  7313538519    finger lakes   4995  2013.0      chevrolet   
390673  7313812494  fredericksburg  14995  2005.0           ford   
399678  7309453096      winchester  21995  2016.0          acura   
400884  7311253070      bellingham   6495  2003.0          honda   
69592   7309745657  visalia-tulare  30590  2019.0          mazda   
48607   7316550263    reno / tahoe   7000  2010.0           ford   
419595  7303581252       la crosse  32990  2011.0  mercedes-benz   
290637  7311128240      cincinnati  15495  2006.0          dodge   
44175   7315536395    palm springs  56800  2015.0           ford   

                              model  condition    cylinders    fuel  odometer  \
422586               silverado 1500  excellent  8 cylinders     gas   28700.0   
269184                        spark       good  4 cylinders     gas  123456.0   
390673                         e150        NaN          NaN     gas   58000.0   
399678                          mdx        NaN          NaN     gas   81876.0   
400884  element ex sport utility 4d        NaN          NaN     gas  159699.0   
69592         cx-5 signature diesel       good          NaN  diesel    7390.0   
48607                    fusion sel       good          NaN     gas   88926.0   
419595             e-class e 63 amg       good  8 cylinders     gas   82763.0   
290637          magnum srt-8 v8 rwd        NaN  8 cylinders     gas  159979.0   
44175                         f-250        NaN          NaN  diesel   64441.0   

        ...                VIN drive       size       type paint_color county  \
422586  ...                NaN   4wd  full-size      truck        grey    NaN   
269184  ...  KL8CD6S91DC505395   fwd        NaN  hatchback       black    NaN   
390673  ...                NaN   NaN        NaN        NaN         NaN    NaN   
399678  ...  5FRYD4H90GB052806   4wd        NaN        SUV       white    NaN   
400884  ...  5J6YH28513L022741   4wd        NaN        SUV        blue    NaN   
69592   ...  JM3KFBE28K0659922   NaN        NaN      other         red    NaN   
48607   ...                NaN   NaN        NaN        NaN         NaN    NaN   
419595  ...  WDDHF7HB1BA274505   rwd        NaN      sedan         NaN    NaN   
290637  ...  2D8GV77316H252547   rwd        NaN      wagon       black    NaN   
44175   ...  1FT7W2BT6FEA34620   4wd        NaN     pickup       black    NaN   

       state        lat        long  posting_date  
422586    wi  42.882100  -88.345100    2021-05-01  
269184    ny  43.225408  -77.348355    2021-04-27  
390673    va  38.157400  -79.065100    2021-04-28  
399678    va  38.762945  -77.460595    2021-04-19  
400884    wa  48.901032 -122.485109    2021-04-23  
69592     ca  36.400000 -119.380000    2021-04-20  
48607     ca  39.640436 -119.711494    2021-05-04  
419595    wi  43.820000  -91.230000    2021-04-08  
290637    oh  39.073100  -84.325896    2021-04-22  
44175     ca  34.107220 -117.455310    2021-05-01  

[10 rows x 22 columns]

### Gestion des doublons

#### Premier dédoublonnage en utilisant l'immatriculation du véhicule 'VIN' et la date du posting 'posting_date'

In [25]:
df.shape

(354200, 22)

In [26]:
# Exclure les lignes sans VIN (on ne veut pas les traiter comme doublons)
df_vin_non_null = df[df['VIN'].notna() & (df['VIN'] != '')]
df_vin_non_null.shape

(222816, 22)

In [27]:
# Dédoublonnage sur (VIN + posting_date) parmi les lignes valides
df_vin_dedoublonne = df_vin_non_null.drop_duplicates(subset=['VIN', 'posting_date'])
df_vin_dedoublonne.shape

(145799, 22)

In [28]:
# Garder les autres lignes non concernées (avec VIN vide ou NaN)
df_sans_vin = df[df['VIN'].isna() | (df['VIN'] == '')]
df_sans_vin.shape

(131384, 22)

In [29]:
# Recombiner les deux
df = pd.concat([df_vin_dedoublonne, df_sans_vin], ignore_index=True)
df.shape

(277183, 22)

#### Deuxième dédoublonnage sur les lignes sans 'VIN' en utilisant les variables importantes et la date du posting 'posting_date'

In [30]:
# Définir les colonnes clés pour le second dédoublonnage
colonnes_importantes = [
    'region', 'price', 'year', 'manufacturer', 'model', 'posting_date'
]
df.shape

(277183, 22)

In [31]:
# Appliquer le drop_duplicates
df = df.drop_duplicates(subset=colonnes_importantes)
df.shape

(275689, 22)

Conclusion du dédoublonnage :
- 77 017 doublons supprimés en utilisant les colonnes 'VIN' + 'Posting_date' (~35%), ce qui montre un très fort taux de repost d'annonce des mêmes voitures dans la même journée
- 1 494 doublons supprimés sur en utilisant les colonnes importantes (~1%)
Cette suppression évite d'avoir des annonces identiques ou trop proches pour augmenter la qualité de nos analyses

## Axes d'analyse

## Vectorisation